### ISSUE TITLE
Support pki file?

### ISSUE THREAD

**Issue by AGou-ops** [kubernetes-sigs/headlamp#2413]
![CleanShot 2024-10-11 at 17 11 46@2x](https://github.com/user-attachments/assets/f24ce43d-648f-4926-99e4-1c40604b4c14)

This kind of kubeconfig is not recognized by headlamp.

**Comment by joaquimrocha** [comment]
Hi @AGou-ops . Can you give us more info about your setup? Mac/Linux/Windows/In-Cluster? If Linux, what Headlamp build have you installed? What is the Headlamp version you are using?

Thanks in advance.

**Comment by AGou-ops** [comment]
@joaquimrocha yes.
- MacOS 14.6.1 m1 pro chip.
- Headlamp 0.25.1

![CleanShot 2024-10-12 at 08 31 58@2x](https://github.com/user-attachments/assets/48aa2baa-323d-4e3b-bb6b-51c4ef7b137c)

If I replace the pki file with base64 data, the cluster will connect properly.

![CleanShot 2024-10-12 at 08 38 54@2x](https://github.com/user-attachments/assets/38c8c6aa-e1be-4936-8f7c-77b60ce0bdf1)

**Comment by joaquimrocha** [comment]
cc/ @knrt10

**Comment by knrt10** [comment]
Let me check. Thank you for reporting

**Comment by knrt10** [comment]
@AGou-ops Do you have access to backend? If so, can you paste logs when you try to use the file directly. I am not sure but I think the path `pki/admin.pem` might be the issue. Headlamp might not be able to load the path. Instead of just pki/admin.pem. Can you try changing it to absolute path from your home directory. Something like `/Users/username/pathtopkifile`? Please try and let me know. Thanks

**Comment by AGou-ops** [comment]
@knrt10 yes. The path and access permissions are both ok, because I can access the cluster normally via kubectl command.

<details>
  <summary> `~/.kube/config` </summary>

```bash
❯ k config view
apiVersion: v1
clusters:
- cluster:
    certificate-authority: pki/ca.pem
    server: https://xxx.xxx.xxx.xxx:16443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: kubernetes-admin
  name: kubernetes-admin@kubernetes
current-context: kubernetes-admin@kubernetes
kind: Config
preferences: {}
users:
- name: kubernetes-admin
  user:
    client-certificate: pki/admin.pem
    client-key: pki/admin.key
❯ k cluster-info
Kubernetes control plane is running at https://xxx.xxx.xxx.xxx:16443
CoreDNS is running at https://xxx.xxx.xxx.xxx:16443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
```

</details>



After changing to an absolute path, it works. So, this relative path config is not supported now?

Thanks.

**Comment by illume** [comment]
@AGou-ops Thanks for testing with the absolute path.

I'm pretty sure this is related to where on Mac Headlamp is not setting some paths as is expected by kubectl libraries.

The HOME or current working directory are suspects (maybe both!).

We have a separate issues about headlamp-plugin not being able to find the HOME plugins folder, so I suspect it's got to do with the HOME. Note, this doesn't happen on all Macs, as I remember seeing several people with Macs where it worked, and a few where it didn't. I can't reproduce it on my mac for example with my config.
- https://github.com/kubernetes-sigs/headlamp/issues/2577


It might help to debug if we log the ENV from headlamp-server.

**Comment by illume** [comment]
I wonder if it’s as simple as changing the current working directory for when loading the config files.

I guess those paths in the original post are relative to ~/.kube/config So maybe to there.

**Comment by danielecr** [comment]
I do not know if already said, there are 2 problems here:
1. the relative path is not solved (the cause of the problem)
2. the error is not properly propagated (it print "Bad Gateway", instead of "Unable to read client certificate. ..." / "... Certificate Authority ...")

First in https://github.com/kubernetes-sigs/headlamp/blob/main/backend/pkg/kubeconfig/kubeconfig.go

The error propagation/detection cause the app trying to connect without key/CA, then print error message Bad Gateway
is probably in cmd/server.go
